### PR TITLE
bug fix and implementation on `denormalization` logic.

### PR DESCRIPTION
**Pull request by jinminhao**
- [x] bug fix on merge_attr
- [x] bug fix on generate and write to csv
- [x] bug fix on denormalziation for continuous field when activating the logp flag 
- [x] add a new field called Word2Vec.

### CODE REVIEW

**Comment by jinminhao** [comment]
I would try to finish the work and close this pr before 1/19. Thanks!

File: netshare/generate/model_generate.py
Comment (jinminhao) [inline netshare/generate/model_generate.py:84]: The logic here is to find the column index which is a *flag* used to represent if the session (flow) start in this chunk.

Previously, the attribute schema is (5-tuple, [session_start_time], *flag*, ....).  Therefore, we need to have a if condition to get the correct index based on if the user specifies `interarrival`(`interarrival` means there is a session_start_time in attributes).

But, after the code refactoring, the schema is (5-tuple, *flag*, ...., , [session_start_time]). Therefore, we don't need this if condition logic anymore.

File: netshare/generate/model_generate.py
Comment (jinminhao) [inline netshare/generate/model_generate.py:115]: Whether we need to concat and merge the attributes is not based on if the user specifies `word2vec_size`. I just deleted that condition here.

File: netshare/generate/model_generate.py
Comment (jinminhao) [inline netshare/generate/model_generate.py:135]: the `row_this_chunk` is constructed but not used in the following logic. I just deleted them.

File: netshare/generate/model_generate.py
Comment (jinminhao) [inline netshare/generate/model_generate.py:169]: This function is no longer in use I guess.

**Comment by sniperyyc** [comment]
> I would try to finish the work and close this pr before 1/19. Thanks!

No rush, please take your time :)

File: netshare/generate/denormalize_fields.py
Comment (jinminhao) [inline netshare/generate/denormalize_fields.py:42]: This may be incorrect. We cannot only get the first dimension on the continuous fields when denormalizing.

File: netshare/generate/model_generate.py
Comment (saart) [reply netshare/generate/model_generate.py:84]: Oh I'm surprised that I changed the schema. 
I'm so sorry for not noticing that!
I hope that the stronger E2E tests that Dorothy currently writing will help make sure that it won't happen again!
Thank you!!

File: netshare/generate/model_generate.py
Comment (saart) [inline netshare/generate/model_generate.py]: I think that no one is using the `word2vec_vecSize` config now, right?
We can drop it 🎉

File: netshare/generate/model_generate.py
Comment (saart) [inline netshare/generate/model_generate.py:94]: your logic depends on the order of the list, and the list is stemmed from a `dict`'s values.
Generally, we shouldn't do that because `dict` is an unordered data structure.
Maybe we should use an OrderedDict here? Alternatively, I'm really not sure why the API stores `dict` and not `list`. We need the dictionary due to the double parsing in `setup_cross_chunks_data:155` and `setup_per_chunk_data:44` (which is really bad and should be removed as well!!), so maybe we can store `list` in the API? I.e. change `learn_api:102`?

File: netshare/generate/model_generate.py
Comment (saart) [inline netshare/generate/model_generate.py:89]:
```suggestion
    for field in learn_api.get_attributes_fields().values():
```
This is more pythonic (and better performance) :)

File: netshare/learn/utils/word2vec_embedding.py
Comment (saart) [inline netshare/learn/utils/word2vec_embedding.py]: Can you use types throughout the file?

File: netshare/learn/utils/word2vec_embedding.py
Comment (saart) [inline netshare/learn/utils/word2vec_embedding.py]: please write an informative log, and it should probably be a debug

File: netshare/generate/denormalize_fields.py
Comment (saart) [inline netshare/generate/denormalize_fields.py]: If that's possible, can you put these lines as a new API in the `learn_api` module, e.g. something like `get_word2vec_model_path`, and make sure that the writing procedure (`word2vec_train`) also uses it?

File: netshare/generate/denormalize_fields.py
Comment (saart) [inline netshare/generate/denormalize_fields.py]: in all these lines you're reading and parsing the learning configuration. It's a code smell that the parsing logic happens twice, in two different components.
Here's a suggestion that will make our life much much easier:
Add the member `is_word2vec_encoding` to `Field` with the default value `False`. 
Populate it during the instance creation in the module `setup_cross_chunks_data` line 163. 
then, here, you can just filter the given `fields_list` using this member.
I.e. this is better because of two big reasons:
1. We encapsulate the configuration parsing logic in a single place
2. you replace 16 lines of complex code (the is a duplication in some sense) with 4 lines of easy code :-)

File: netshare/generate/denormalize_fields.py
Comment (saart) [inline netshare/generate/denormalize_fields.py]:
```suggestion
            for k in word2vec_dict_type_cols:
```
Just a syntax sugar :-)

File: netshare/generate/denormalize_fields.py
Comment (saart) [reply netshare/generate/denormalize_fields.py]: Another benefit will be that the rest of your logic will use the `Field` object, and not `Any`, such as this function signature: `get_word2vec_type_col(word2vec_cols: List[Any])`

File: netshare/learn/utils/word2vec_embedding.py
Comment (saart) [inline netshare/learn/utils/word2vec_embedding.py]: The type of this structure is pretty complex. It is pretty hard to understand how to use it. e.g. when you wrote `dict_type_annDictPair[word2vec_type_col][0]` it is pretty hard to follow, and this is a source of bugs.
WDYT about using `namedtuple` here?
Here's a suggestion:
```
from typing import NamedTuple

class annoyTypeDescription(NamedTuple):
   annoy_type: int
   annoy_dict: Dict[int, ???]  # I really don't know
```
and then your return value will be: `Dict[str, annoyTypeDescription]`

File: netshare/generate/model_generate.py
Comment (saart) [reply netshare/generate/model_generate.py:94]: If it works for you, then we can keep it like that, and I'll change it after the big merge !

File: netshare/generate/model_generate.py
Comment (jinminhao) [reply netshare/generate/model_generate.py]: Yes, `word2vec_vecSize` is no longer usable.

File: netshare/generate/model_generate.py
Comment (jinminhao) [reply netshare/generate/model_generate.py:94]: I agree. `List` or `OrderedDict` is definitely better. I also have this concern and when I'm iterating the `dict`, I find it out the order is correct. So I just plan to put it there and discuss it with you. I agree that using list is better, I will make the change on the code. Thanks!!

File: netshare/generate/denormalize_fields.py
Comment (jinminhao) [reply netshare/generate/denormalize_fields.py]: Sure. I will also change a bit on `word2vec_train`

File: netshare/generate/denormalize_fields.py
Comment (jinminhao) [reply netshare/generate/denormalize_fields.py]: Yes, I agree that adding word2vec flag into field is a great idea. Previously, I also had a chat with @sniperyyc on this stuff. I'm not doing that because the modification may include a lot of files. I'm bit afraid to have a large conflit with you on the branch `code-refactor` if you have some continue update on that. Since the work on `code-refactor` is almost done, I can do that!

File: netshare/learn/utils/word2vec_embedding.py
Comment (jinminhao) [reply netshare/learn/utils/word2vec_embedding.py]: I think so. I can change it.

File: netshare/generate/model_generate.py
Comment (jinminhao) [reply netshare/generate/model_generate.py:84]: No wories. It's because `apply_timestamp_generation` being called in `apply_distributed_chunk_logic` is after `build_fields` being called in `setup_cross_chunks_data`

File: netshare/generate/model_generate.py
Comment (saart) [reply netshare/generate/model_generate.py:94]: Great, thank you 🙏 
If you prefer, you can leave it like that and I'll fix it in a followup PR

File: netshare/generate/denormalize_fields.py
Comment (saart) [reply netshare/generate/denormalize_fields.py]: I don't have any work left in `code-refactor`, so don't worry about more conflicts :)

Comment (sniperyyc) [reply netshare/generate/denormalize_fields.py:62]: > 

Yes, I remembered our discussion. Adding a `word2vec` field is definitely a more generic/long-term solution to this.

Comment (sniperyyc) [reply netshare/generate/denormalize_fields.py]: > Yes, I agree that adding word2vec flag into field is a great idea. Previously, I also had a chat with @sniperyyc on this stuff. I'm not doing that because the modification may include a lot of files. I'm bit afraid to have a large conflit with you on the branch `code-refactor` if you have some continue update on that. Since the work on `code-refactor` is almost done, I can do that!

Yes, I remembered our discussion. Adding a `word2vec` field is definitely a more generic/long-term solution to this.

File: netshare/generate/model_generate.py
Comment (jinminhao) [reply netshare/generate/model_generate.py:94]: Sure. I will leave it to you.

File: netshare/generate/model_generate.py
Comment (saart) [reply netshare/generate/model_generate.py:84]: Ohhh I see, wow, so sorry for that :(
If you prefer, we can also move the logic, or change the place where we put the new fields (i.e. the line `new_timeseries_list.insert(0, time_col)`)...
Amazing that you spot it 👑

File: netshare/utils/field.py
Comment (saart) [inline netshare/utils/field.py:218]: can you add typing here?
Maybe I'll add it as a constraint in mypy on my next PR :)

File: netshare/utils/word2vec_utils.py
Comment (saart) [inline netshare/utils/word2vec_utils.py]: I think that this is just an empty file?

File: netshare/learn/setup_cross_chunks_data.py
Comment (saart) [inline netshare/learn/setup_cross_chunks_data.py:128]: I thought that `build_annoy_dictionary_word2vec` will call this function itself, and not get it as an input,
But that's really not mandatory

File: netshare/learn/utils/word2vec_embedding.py
Comment (saart) [inline netshare/learn/utils/word2vec_embedding.py]: Can you please delete?

File: netshare/learn/utils/word2vec_embedding.py
Comment (saart) [inline netshare/learn/utils/word2vec_embedding.py]:
```suggestion
        for index, ele in enumerate(type_set):
            type_ann.add_item(index, get_vector(model, str(ele), norm_option=True))
            type_dict[index] = ele
```
Syntax sugar :)

Comment (saart) [reply netshare/learn/utils/word2vec_embedding.py]: don't you prefer that `type_dict` will be a list?

File: netshare/learn/utils/word2vec_embedding.py
Comment (saart) [inline netshare/learn/utils/word2vec_embedding.py]: Can you add types to all the functions in this file?

File: netshare/utils/field.py
Comment (saart) [inline netshare/utils/field.py]:
```suggestion
        for type in dict_annDictPair:
```

File: netshare/utils/field.py
Comment (saart) [inline netshare/utils/field.py]: you might get here `AttributeError`. 
You probably would have found it if your wrote tests - but it's not mandatory, so it's your choice :)

Comment (saart) [reply netshare/utils/field.py]: btw, a pretty cool feature that does what you're trying to do here is `for...else`
Google it :)

File: netshare/learn/utils/word2vec_embedding.py
Comment (jinminhao) [reply netshare/learn/utils/word2vec_embedding.py]: `type_dict`, is used to help to find out the nearest 'word' when denormalizing. Therefore, using dict for fast search for the correct 'word' may be important..

File: netshare/utils/word2vec_utils.py
Comment (jinminhao) [reply netshare/utils/word2vec_utils.py]: Yes, sorry about that:)

File: netshare/learn/utils/word2vec_embedding.py
Comment (jinminhao) [reply netshare/learn/utils/word2vec_embedding.py]: Sure

File: netshare/generate/denormalize_fields.py
Comment (jinminhao) [inline netshare/generate/denormalize_fields.py]: When writing to csv, we need to rely on the gen flag to check if the session has already ended at a specific time step.

File: netshare/utils/field.py
Comment (jinminhao) [inline netshare/utils/field.py:70]: When activating log1p_norm, we need also to norm the min and max. Because in denoamlizing stage, the generated data will first be denomalized by min-max denoamlizagion, then expm1. Using a unnormalized min max may made the result larger then expected and sometimes may cause it become inf.

File: netshare/generate/denormalize_fields.py
Comment (saart) [inline netshare/generate/denormalize_fields.py]:
```suggestion
        if is_session_key and len(sub_data.shape) == 1:
```
Search for "greeting" in https://peps.python.org/pep-0008/

File: netshare/generate/denormalize_fields.py
Comment (saart) [inline netshare/generate/denormalize_fields.py]:
```suggestion
        if not is_session_key and len(sub_data.shape) == 2:
```

File: netshare/generate/denormalize_fields.py
Comment (saart) [reply netshare/generate/denormalize_fields.py]: (absolutely nit-picking, you can ignore if you wish)

File: netshare/generate/denormalize_fields.py
Comment (saart) [inline netshare/generate/denormalize_fields.py]:
```suggestion
        for data_gen, session_data, timeserie in zip(data_gen_flag, session_key, timeseries):
```
And then you can forget about the index `i`.
(Please check it out - I'm not sure that it will work with np.arrays, but if it doesn't - it helps for readability)

File: netshare/learn/setup_cross_chunks_data.py
Comment (saart) [inline netshare/learn/setup_cross_chunks_data.py]:
```suggestion
) -> Tuple[Optional[Word2Vec], Optional[List[str]]]:
```
maybe?

File: netshare/learn/utils/word2vec_embedding.py
Comment (saart) [inline netshare/learn/utils/word2vec_embedding.py]:
```suggestion
    dict_type_cols: Dict[str, List[str]] = defaultdict(list)
    for col in word2vec_cols:
        dict_type_cols[col.encoding].append(col.column)
    return dict_type_cols
```
`from collections import defaultdict` This is one of the best utilities :)

File: netshare/learn/utils/word2vec_embedding.py
Comment (saart) [inline netshare/learn/utils/word2vec_embedding.py:74]: so the type is structured from two different entities? 
`word2vec` which tells us the type, and `port`/`proto` which tells us the "group of words" that it should be included in?
Why isn't that another key in the field's configuration, and just maintain it as a member of `Word2VecField`?

Comment (saart) [reply netshare/learn/utils/word2vec_embedding.py:74]: And then, you can use `word2vec_fields: List[Field]` instead of `word2vec_cols: List[Any]` everywhere.

Comment (saart) [reply netshare/learn/utils/word2vec_embedding.py:74]: But: Maybe this is not in the scope of this PR, and we should do this change later.
It's your choice :)

File: netshare/learn/utils/word2vec_embedding.py
Comment (saart) [inline netshare/learn/utils/word2vec_embedding.py]:
```suggestion
          itertools.chain.from_iterable([list(df[col]) for col in cols])
```
Much faster :-)

File: netshare/learn/utils/word2vec_embedding.py
Comment (saart) [inline netshare/learn/utils/word2vec_embedding.py]:
```suggestion
    return [get_original_obj(ann, vector, dic) for vector in vectors]
```
DRY
right?

File: netshare/learn/utils/word2vec_embedding.py
Comment (saart) [inline netshare/learn/utils/word2vec_embedding.py]: delete

File: netshare/utils/field.py
Comment (saart) [reply netshare/utils/field.py:70]: That's a good catch!
Can you please move this code to the __init__?
`normalize` should be a pure function: I.e. we don't expect it to change any internal value while operating. I.e. someone can call it multiple times and that's fine. Here, you're changing members inside `self`

File: netshare/utils/field.py
Comment (saart) [inline netshare/utils/field.py:250]: sounds like an awful bug 😂 
Well done!!!
Here's a relevant link: https://stackoverflow.com/questions/1450957/pythons-json-module-converts-int-dictionary-keys-to-strings

**Review by saart — APPROVED** [review]
LGTM!
Well done!!

File: netshare/generate/denormalize_fields.py
Comment (jinminhao) [reply netshare/generate/denormalize_fields.py]: It's definitely more elegent. Thanks!!

File: netshare/learn/utils/word2vec_embedding.py
Comment (jinminhao) [reply netshare/learn/utils/word2vec_embedding.py:74]: Great idea! i think so. Let me mark it and put it into another PR

**Comment by jinminhao** [comment]
> LGTM! Well done!!

Thank you so much for your review and comments!!! I just gave a final update and will merge it.

**Pull request merged**
